File: NewVM.fs
// Create an azure-native.compute.VirtualMachine Domain Controller server in F#
// https://www.pulumi.com/ai/conversations/13fb300d-c234-4b22-9b07-84d2139c876b

open Pulumi
open Pulumi.AzureNative.Compute
open Pulumi.AzureNative.Compute.Inputs
open Pulumi.AzureNative.Network
open Pulumi.AzureNative.Network.Inputs
open Pulumi.AzureNative.Resources

let infra () =
    task {
        // Create a resource group
        let! resourceGroup = ResourceGroup("adResourceGroup")

        // Create a virtual network
        let! virtualNetwork = VirtualNetwork("adVnet",
                                   VirtualNetworkArgs(
                                       ResourceGroupName = input resourceGroup.Name,
                                       AddressSpaces = inputList [ "10.0.0.0/16" ]))

        // Create a subnet within the virtual network
        let! subnet = Subnet("adSubnet",
                             SubnetArgs(
                                 ResourceGroupName = input resourceGroup.Name,
                                 VirtualNetworkName = input virtualNetwork.Name,
                                 AddressPrefix = "10.0.1.0/24"))

        // Create a public IP address
        let! publicIPAddress = PublicIPAddress("adPublicIP",
                                               PublicIPAddressArgs(
                                                   ResourceGroupName = input resourceGroup.Name,
                                                   PublicIPAllocationMethod = input PublicIPAllocationMethod.Static))

        // Create a network interface
        let! networkInterface = NetworkInterface("adNic",
                                                 NetworkInterfaceArgs(
                                                     ResourceGroupName = input resourceGroup.Name,
                                                     IpConfigurations = inputList [
                                                         NetworkInterfaceIPConfigurationArgs(
                                                             Name = "adNicIPConfig",
                                                             Subnet = input subnet.GetResource().ToReference(),
                                                             PrivateIPAllocationMethod = input PrivateIPAllocationMethod.Dynamic,
                                                             PublicIPAddress = input publicIPAddress.GetResource().ToReference())
                                                     ]))

        // Define the storage profile
        let storageProfile = StorageProfileArgs(
            OsDisk = OSDiskArgs(
                Caching = input CachingTypes.ReadWrite,
                ManagedDisk = ManagedDiskParametersArgs(
                    StorageAccountType = StorageAccountTypes.Standard_LRS),
                Name = "adOsDisk",
                CreateOption = input DiskCreateOptionTypes.FromImage),
            ImageReference = ImageReferenceArgs(
                Publisher = "MicrosoftWindowsServer",
                Offer = "WindowsServer",
                Sku = "2019-Datacenter",
                Version = "latest"))

        // Define the OS profile
        let osProfile = OSProfileArgs(
            ComputerName = "adVM",
            AdminUsername = "admin-user",
            AdminPassword = input "P@ssw0rd!",
            WindowsConfiguration = WindowsConfigurationArgs(
                ProvisionVMAgent = input true,
                EnableAutomaticUpdates = input true))

        // Define the network profile
        let networkProfile = NetworkProfileArgs(
            NetworkInterfaces = inputList [ NetworkInterfaceReferenceArgs(Id = input networkInterface.Id) ])

        // Create the virtual machine
        let! vm = VirtualMachine("adVM",
                                 VirtualMachineArgs(
                                     ResourceGroupName = input resourceGroup.Name,
                                     NetworkProfile = networkProfile,
                                     OsProfile = osProfile,
                                     StorageProfile = storageProfile,
                                     HardwareProfile = HardwareProfileArgs(VmSize = VirtualMachineSizeTypes.Standard_DS2_v2),
                                     Location = input resourceGroup.Location))

        return {
            Pulumi.UserName = vm.Id
        }
    }

[<EntryPoint>]
let main _ =
    Deployment.run infra

File: NewVM.cs
// Create an azure-native.compute.VirtualMachine Domain Controller server
// https://www.pulumi.com/ai/conversations/d60bac32-0d8f-40e3-b18e-d3c48bca1f45

using Pulumi;
using Pulumi.AzureNative.Compute;
using Pulumi.AzureNative.Compute.Inputs;
using Pulumi.AzureNative.Network;
using Pulumi.AzureNative.Network.Inputs;
using Pulumi.AzureNative.Resources;
using System.Collections.Generic;

return await Deployment.RunAsync(() =>
{
    // Create an Azure Resource Group
    var resourceGroup = new ResourceGroup("domainControllerResourceGroup");

    // Create a Virtual Network
    var virtualNetwork = new VirtualNetwork("domainControllerVNet", new VirtualNetworkArgs
    {
        ResourceGroupName = resourceGroup.Name,
        AddressSpace = new AddressSpaceArgs
        {
            AddressPrefixes = { "10.0.0.0/16" }
        }
    });

    // Create a Subnet
    var subnet = new Subnet("domainControllerSubnet", new SubnetArgs
    {
        ResourceGroupName = resourceGroup.Name,
        VirtualNetworkName = virtualNetwork.Name,
        AddressPrefix = "10.0.1.0/24"
    });

    // Create a Public IP Address
    var publicIp = new PublicIPAddress("domainControllerPublicIp", new PublicIPAddressArgs
    {
        ResourceGroupName = resourceGroup.Name,
        PublicIPAllocationMethod = "Dynamic"
    });

    // Create a Network Interface
    var networkInterface = new NetworkInterface("domainControllerNic", new NetworkInterfaceArgs
    {
        ResourceGroupName = resourceGroup.Name,
        IpConfigurations =
        {
            new NetworkInterfaceIpConfigurationArgs
            {
                Name = "primary",
                Primary = true,
                Subnet = new SubnetArgs
                {
                    Id = subnet.Id,
                },
                PrivateIPAllocationMethod = "Dynamic",
                PublicIPAddress = new PublicIPAddressArgs
                {
                    Id = publicIp.Id,
                }
            }
        }
    });

    // Create a Virtual Machine
    var virtualMachine = new VirtualMachine("domainControllerVM", new VirtualMachineArgs
    {
        ResourceGroupName = resourceGroup.Name,
        Location = resourceGroup.Location,
        NetworkProfile = new NetworkProfileArgs
        {
            NetworkInterfaces =
            {
                new NetworkInterfaceReferenceArgs
                {
                    Id = networkInterface.Id,
                    Primary = true,
                }
            }
        },
        HardwareProfile = new HardwareProfileArgs
        {
            VmSize = "Standard_B2s",
        },
        OsProfile = new OsProfileArgs
        {
            AdminUsername = "adminuser",
            AdminPassword = "Password12345!",
            ComputerName = "domaincontroller"
        },
        StorageProfile = new StorageProfileArgs
        {
            OsDisk = new OSDiskArgs
            {
                CreateOption = DiskCreateOption.FromImage,
                ManagedDisk = new ManagedDiskParametersArgs
                {
                    StorageAccountType = "Standard_LRS"
                }
            },
            ImageReference = new ImageReferenceArgs
            {
                Offer = "WindowsServer",
                Publisher = "MicrosoftWindowsServer",
                Sku = "2019-Datacenter",
                Version = "latest"
            }
        }
    });

    return new Dictionary<string, object?>
    {
        ["vmName"] = virtualMachine.Name,
        ["publicIpAddress"] = publicIp.IpAddress
    };
});

